src/Dependencies/Collections/SegmentedList`1.cs

```diff
@@ -135,12 +135,39 @@ public int Capacity
                 {
                     if (value > 0)
                     {
-                        var newItems = new SegmentedArray<T>(value);
-                        if (_size > 0)
+                        // Rather than creating a copy of _items, instead reuse as much of it's data as possible.
+                        // This saves as much as 50% of allocations and 70% of CPU cost of Add in large collections.
+                        // See SegmentedListBenchmarks_Add for repro details.
+                        var segments = SegmentedCollectionsMarshal.AsSegments(_items);
+
+                        var segmentSize = SegmentedArrayHelper.GetSegmentSize<T>();
+                        var segmentShift = SegmentedArrayHelper.GetSegmentShift<T>();
```

Comment:
Inlining these values will likely improve performance. They are meticulously JIT-optimized.

Reply:
Dustin just explained this, very cool!

Reply:
➡️ Now resolved